baseapp/auth/saml/serviceprovider.go

```diff
@@ -0,0 +1,218 @@
+}
+
+func newSAMLError(err error, status int) samlError {
+	return samlError{
+		Err:          err,
+		ResponseCode: status,
+	}
+}
+
+// ErrorCallback is called whenever an error occurs in the saml package.
+// The callback is expected to send a response to the request. The http.ResponseWriter
+// will not have been written to, allowing the callback to send headers if desired.
+type ErrorCallback func(http.ResponseWriter, *http.Request, samlError)
+
+// LoginCallback is called whenever an auth flow is successfully completed.
+// The callback is responsible preserving the login state.
+type LoginCallback func(http.ResponseWriter, *http.Request, *saml.Assertion)
+
+type Settings struct {
+	AssertionConsumerServicePath string
+	SPMetadataPath               string
+
+	// Whether or not all generated urls should use https.
+	// If false the generated urls will match the incoming request
+	ForceTLS    bool
+	Certificate *x509.Certificate
+	Key         *rsa.PrivateKey
+	IDPMetadata *saml.EntityDescriptor
+	OnError     ErrorCallback
+	OnLogin     LoginCallback
+	IDStore     IDStore
+}
+
+// ServiceProvider is capable of handling a SAML login. It provides
+// an http.Handler (via ACSHandler) which can process the http POST from the SAML IDP. It accepts callbacks for both error and
+// success conditions so that clients can take action after the auth flow is complete. It also provides a handler
+// for serving the service provider metadata XML.
+type ServiceProvider struct {
+	sp           *saml.ServiceProvider
+	acsPath      string
+	metadataPath string
+	forceTLS     bool
+	onError      ErrorCallback
+	onLogin      LoginCallback
+	idStore      IDStore
+}
+
+// NewServiceProviderFromMetadata returns a ServiceProvider. The configuration of the ServiceProvider
+// is a result of combinging settings provided to this method and values parsed from the IDP's metadata.
+func NewServiceProviderFromMetadata(settings Settings) (*ServiceProvider, error) {
+	if settings.Certificate == nil || settings.Key == nil {
+		return nil, errors.New("a certificate and key must be provided")
+	}
+
+	if settings.IDPMetadata == nil {
+		return nil, errors.New("the IDP Metadata must be provided")
+	}
+
+	provider := &saml.ServiceProvider{
+		IDPMetadata: settings.IDPMetadata,
+		Certificate: settings.Certificate,
+		Key:         settings.Key,
+	}
+	sp := &ServiceProvider{
+		sp:           provider,
+		forceTLS:     settings.ForceTLS,
+		onError:      settings.OnError,
+		onLogin:      settings.OnLogin,
+		acsPath:      settings.AssertionConsumerServicePath,
+		metadataPath: settings.SPMetadataPath,
+		idStore:      settings.IDStore,
+	}
+
+	if sp.onError == nil {
+		sp.onError = DefaultErrorCallback
+	}
+
+	if sp.onLogin == nil {
+		sp.onLogin = DefaultLoginCallback
+	}
+
+	if sp.idStore == nil {
+		sp.idStore = cookieIDStore{}
+	}
+
+	return sp, nil
+}
+
+func DefaultErrorCallback(w http.ResponseWriter, r *http.Request, err samlError) {
+	fmt.Println(err.Error())
```

Comment:
Libraries that print always make me nervous... since the default handler is pretty simple, can we remove the print and rely on clients to customize the handler with their preferred logging method?